### PR TITLE
[wallet] wallet-tool: command to remove key metadata

### PR DESCRIPTION
**Pull request by Sjors**
Adds a `keymetadata` method to `wallet-tool` with currently only a `delete` operation.

### CODE REVIEW

**Comment by Sjors** [comment]
I found this useful for salvaging wallets touched by intermediate versions of #14021.

It is a bit dangerous, so perhaps some seatbelt is useful? E.g. we could disable this command for mainnet in release builds.

A useful followup would be a `reconstruct` operation, which takes the master key, re-derives all keys, finds those in the wallet and then restores key origin info. It could take an additional version argument, which restores key metadata for earlier wallet versions. The latter can be useful for backwards compatibility and upgrade path testing.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -358,6 +358,20 @@ void CWallet::LoadScriptMetadata(const CScriptID& script_id, const CKeyMetadata
     m_script_metadata[script_id] = meta;
 }
 
+bool CWallet::EraseWatchOnlyMeta(const CScript& script)
+{
+    AssertLockHeld(cs_wallet); // m_script_metadata
```
Comment (practicalswift) [inline src/wallet/wallet.cpp]: Could drop this: we already have the locking annotations.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -399,6 +413,37 @@ void CWallet::UpgradeKeyMetadata()
     SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA);
 }
 
+void CWallet::DeleteKeyMetadata()
+{
+    AssertLockHeld(cs_wallet);
```
Comment (practicalswift) [inline src/wallet/wallet.cpp]: Could drop this: we already have the locking annotations.

File: src/bitcoin-wallet.cpp [hunk 3]
```diff
@@ -100,18 +102,56 @@ int main(int argc, char* argv[])
...
+            return EXIT_FAILURE;
+        }
+        const std::set<std::string> operations = {
+            "delete"
+        };
+        if (arguments.empty() || operations.find(arguments[0]) == operations.end()) {
+            fprintf(stderr, "Specify operation: delete\n");
+            return EXIT_FAILURE;
+        } else {
+            if (!WalletTool::ExecuteKeyMetadata(name, arguments[0])) return EXIT_FAILURE;
```
Comment (practicalswift) [inline src/bitcoin-wallet.cpp]: Else after return: could drop `else` here?

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -897,6 +901,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
 
     //! Adds a watch-only address to the store, and saves it to disk.
     bool AddWatchOnly(const CScript& dest, int64_t nCreateTime) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    bool EraseWatchOnlyMeta(const CScript &dest) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Comment (practicalswift) [inline src/wallet/wallet.h]: Make sure parameter names match between definition and declaration :-)

File: src/wallet/wallettool.cpp [hunk 5]
```diff
@@ -68,23 +68,26 @@ static std::shared_ptr<CWallet> LoadWallet(const std::string& name, const fs::pa
     }
```
Comment (practicalswift) [inline src/wallet/wallettool.cpp]: When changing this file: please consider changing to const reference in `catch` on L65.

File: test/functional/tool_wallet.py [hunk 6]
```diff
@@ -97,5 +101,32 @@ def run_test(self):
         assert_equal(1000, out['keypoolsize_hd_internal'])
         assert_equal(True, 'hdseedid' in out)
 
+        self.log.info("Test method: keymetadata")
+        self.start_node(0, ['-wallet=foo'])
+        address = self.nodes[0].getnewaddress()
+        address_info_before = self.nodes[0].getaddressinfo(address)
+        assert("hdmasterfingerprint" in address_info_before)
+        self.nodes[0].importmulti([{"desc": "wpkh([deadbeef/1/2'/3/4']03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd)",
+                               "timestamp": "now",
```
Comment (practicalswift) [inline test/functional/tool_wallet.py]: Weird indentation: Consider using `black` to get the proper indentation :-)

**Comment by Sjors** [comment]
Closing for now due to lack of interest and me not using it much either.

**Comment by laanwj** [comment]
> It is a bit dangerous, so perhaps some seatbelt is useful? E.g. we could disable this command for mainnet in release builds.

Don't know whether this kind of very case-specific recovery tooling belongs in `wallet-tool` or makes more sense as a separate script/tool for direct wallet database manipulation. I think in general it's very hard to keep maintained because only few people will have interest in it at a certain time, but when you do happen to need it it's great to have.

**Pull request closed without merging**